src/Workspaces/Core/Portable/EmbeddedLanguages/EmbeddedLanguageDetector.cs

```diff
@@ -0,0 +1,360 @@
+
+            return false;
+        }
+
+        private bool HasLanguageComment(
+            SyntaxTriviaList list,
+            ISyntaxFacts syntaxFacts,
+            [NotNullWhen(true)] out string? identifier,
+            [NotNullWhen(true)] out IEnumerable<string>? options)
+        {
+            foreach (var trivia in list)
+            {
+                if (HasLanguageComment(trivia, syntaxFacts, out identifier, out options))
+                    return true;
+            }
+
+            identifier = null;
+            options = null;
+            return false;
+        }
+
+        private bool HasLanguageComment(
+            SyntaxTrivia trivia,
+            ISyntaxFacts syntaxFacts,
+            [NotNullWhen(true)] out string? identifier,
+            [NotNullWhen(true)] out IEnumerable<string>? options)
+        {
+            if (syntaxFacts.IsRegularComment(trivia))
+            {
+                // Note: ToString on SyntaxTrivia is non-allocating.  It will just return the
+                // underlying text that the trivia is already pointing to.
+                var text = trivia.ToString();
+                if (_commentDetector.TryMatch(text, out identifier, out options))
+                    return true;
+            }
+
+            identifier = null;
+            options = null;
+            return false;
+        }
+
+        private bool IsEmbeddedLanguageInterpolatedStringTextToken(
+            SyntaxToken token,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken,
+            [NotNullWhen(true)] out string? identifier)
+        {
+            // If we have the format string for an interpolation (e.g. `{expr:XXX}`) then see if expr has an
+            // implementation if IFormattable.Format(string, ...) and then see if that impl method has a
+            // [StringSyntax] attribute on the first parameter.
+
+            identifier = null;
+            var syntaxFacts = Info.SyntaxFacts;
+            var interpolationFormatClause = token.Parent;
+            var interpolation = interpolationFormatClause?.Parent;
+            if (interpolation?.RawKind != syntaxFacts.SyntaxKinds.Interpolation)
+                return false;
+
+            var expression = syntaxFacts.GetExpressionOfInterpolation(interpolation);
+            var type = semanticModel.GetTypeInfo(expression, cancellationToken).Type;
+            if (type == null)
+                return false;
+
+            var iformattable = type.AllInterfaces.FirstOrDefault(t => t is
+            {
+                Name: nameof(IFormattable),
+                ContainingNamespace:
+                {
+                    Name: nameof(System),
+                    ContainingNamespace.IsGlobalNamespace: true,
+                }
+            });
+            if (iformattable == null)
+                return false;
+
+            var formatMethod = iformattable
+                .GetMembers(nameof(IFormattable.ToString))
```

Comment:
What about all [the methods](https://docs.microsoft.com/dotnet/standard/base-types/composite-formatting) that do not meet this pattern? Presumably we would want this to work on older frameworks?

Reply:
this is for interpolated strings (e.g. `{foo:XXX}`).  So this is the method that the runtime will be calling through.  if there are other methods that we actually need because users run into them, we can def add it.